### PR TITLE
Task 12: Updated selectors to make them more stable and role-agnostic

### PR DESCRIPTION
**Pull request by vpostolov2**
Updated selectors to make them more stable and role-agnostic

### CODE REVIEW

File: src/main/java/cz/czechitas/automation/ApplicationAction.java [hunk 1]
```diff
@@ -18,42 +18,42 @@ final class ApplicationAction {
...
+     */
     void clickCreateNewApplicationButton() {
-        var createApplicationButton = elementFinder.findByXPath("/html/body/div/div/div/div/div/div[1]/a");
+        var createApplicationButton = elementFinder.findByCssSelector(".card-header a");
         createApplicationButton.click();
     }
 
     void selectProgrammingSection() {
-        var programmingSectionButton = elementFinder.findByXPath("/html/body/div/div/div[1]/div[3]/div/div[2]/a");
+        var programmingSectionButton = elementFinder.findByXPath("//*[contains(text(), 'Programování')]/../..//a");
```
Comment (N0081K) [inline src/main/java/cz/czechitas/automation/ApplicationAction.java:30]:
```suggestion
        var programmingSectionButton = elementFinder.findByXPath("//*[contains(text(), 'Programování')]//ancestor::*[@class='card']//a");
```

My personal preference is to not rely on absolute paths to ancestors 🙂

File: src/main/java/cz/czechitas/automation/LoginAction.java [hunk 2]
```diff
@@ -20,33 +20,33 @@ final class LoginAction {
...
 
     void clickLoginButton() {
-        var loginButton = elementFinder.findByXPath("/html/body/div/div/div/div/div/div/form/div[3]/div/button");
+        var loginButton = elementFinder.findByXPath("//button[@type='submit']");
         loginButton.click();
     }
 
     void logout() {
-        var signedInUserElement = elementFinder.findByXPath("//*[@id=\"navbarSupportedContent\"]/div[2]/div/a");
+        var signedInUserElement = elementFinder.findByXPath("//*[@id='navbarSupportedContent']/div[2]/div/a");
```
Comment (N0081K) [inline src/main/java/cz/czechitas/automation/LoginAction.java:47]: I would not rely that much on the order 🙂 
```suggestion
        var signedInUserElement = elementFinder.findByXPath("//*[@id='navbarSupportedContent']//a[@class='dropdown-toggle']");
```

**Review by N0081K — COMMENTED** [review]
overall a lot of good work. Sometimes it is impossible here to get good selectors.
I noticed some that won't work always 2 of them.
And 3 more that uses pattern we try to avoid at all cost in my job (but we do not have to replace them

**Comment by shinedeveloper** [comment]
The best approach to selectors is in the first place contact Pavel Kutac, the author of the app under test and add there semantic data- aatributes or classes so that we don't have to have these awkward selectors.

**Comment by shinedeveloper** [comment]
Even better, prepare right away a pull request for the Pavel Kutac's  app under test with the improvements.

**Comment by N0081K** [comment]
True, the repo is here https://github.com/czechitas/da-test-webapp

File: src/main/java/cz/czechitas/automation/ApplicationAction.java [hunk 3]
```diff
@@ -18,42 +18,42 @@ final class ApplicationAction {
...
         var searchInput = elementFinder.findByXPath(
-                "/html/body/div/div/div/div/div/div[2]/div[1]/div/div[2]/div/label/input");
+                "//input[@type='search']");
         searchInput.sendKeys(textToSearch);
     }
 
     void clickEditFirstApplicationButton() {
         var firstApplicationEditButton = elementFinder.findByXPath(
-                "/html/body/div/div/div/div/div/div[2]/div[2]/div/table/tbody/tr[1]/td[5]/div/a[2]");
+                "//a[contains(text(), 'Upravit')]");
```
[the remark this answers is not in the document]
Comment (vpostolov2) [reply src/main/java/cz/czechitas/automation/ApplicationAction.java:56]: Fixed for Upravit and Zobrazit.

File: src/main/java/cz/czechitas/automation/LoginAction.java (hunk 2 above)
Comment (vpostolov2) [reply src/main/java/cz/czechitas/automation/LoginAction.java:47]: Fixed

File: src/main/java/cz/czechitas/automation/ApplicationAction.java [hunk 4]
```diff
@@ -18,42 +18,42 @@ final class ApplicationAction {
...
     void clickCreateApplicationButton() {
-        var createApplicationButton = elementFinder.findByXPath("/html/body/div/div/div/div/div/div/div/div[2]/a");
+        var createApplicationButton = elementFinder.findByCssSelector(".card-body a");
         createApplicationButton.click();
     }
 
     void openFirstApplicationDetailsPage() {
         var openApplicationButton = elementFinder.findByXPath(
-                "/html/body/div/div/div/div/div/div[2]/div[2]/div/table/tbody/tr/td[5]/div/a[1]");
+                "//a[contains(text(), 'Detail')]");
```
[the remark this answers is not in the document]
Comment (vpostolov2) [reply src/main/java/cz/czechitas/automation/ApplicationAction.java:44]: Fixed

**Comment by vpostolov2** [comment]
I think selectors are more or less fine, most of them are really simple and understandable. Also, ladies don't work within the framework itself on mandatory lectures, so they won't struggle too.

@N0081K please take another look once you have a chance. Thanks!

**Review by N0081K — APPROVED** [review]

**Pull request merged**